pkg/controllers/management/globaldns/globaldns_handler.go

```diff
@@ -0,0 +1,228 @@
+package globaldns
+
+import (
+	"context"
+	"net"
+	"sort"
+
+	v1beta1Rancher "github.com/rancher/types/apis/extensions/v1beta1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	clientv1beta1 "k8s.io/client-go/kubernetes/typed/extensions/v1beta1"
+)
+
+const (
+	GlobaldnsController = "mgmt-global-dns-controller"
+)
+
+type GlobalDNSController struct {
+	globalDNSInterface v3.GlobalDNSInterface
+	globalDNSLister    v3.GlobalDNSLister
+	//ingressInterface   clientv1beta1.IngressInterface //need to use client-go IngressInterface to update Ingress.Status field
+	ingressLister     v1beta1Rancher.IngressLister
+	ctx               context.Context
+	managementContext *config.ManagementContext
+}
+
+func newGlobalDNSController(ctx context.Context, mgmt *config.ManagementContext) *GlobalDNSController {
+	extInterface, err := v1beta1Rancher.NewForConfig(mgmt.RESTConfig)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	n := &GlobalDNSController{
+		globalDNSInterface: mgmt.Management.GlobalDNSs(""),
+		globalDNSLister:    mgmt.Management.GlobalDNSs("").Controller().Lister(),
+		ctx:                ctx,
+		ingressLister:      extInterface.Ingresses("default").Controller().Lister(),
+		managementContext:  mgmt,
+	}
+	return n
+}
+
+func (n *GlobalDNSController) getIngressInterface(namespace string) clientv1beta1.IngressInterface {
+	if namespace != "" {
+		return n.managementContext.K8sClient.Extensions().Ingresses(namespace)
+	}
+	return nil
```

Comment:
we never check for nil in the callers, can we either drop the check for the namespaces, or add nil handling?

Reply:
Or better lets add Ingresses client directly to the struct, and access it w/o the getter

Reply:
done